### PR TITLE
Run all fallback kernels with FakeTensor

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88248

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10

This improves the memory compression of resnet18 from .84 -> .94 on inductor no-cudagraphs. It does mean that any extern kernel which incorrectly computes strides will be a hard error at runtime, but that's an issue we are going to have to face with dynamic shapes anyway. CC @ezyang, @SherlockNoMad

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
I'm surprised no conversions are necessary

**Comment by eellison** [comment]
@ezyang the inputs are TorchInductor IR, not real tensors, and within process_kernel we do the instantiations of real tensors. Thus no conversion needed.

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -309,6 +309,13 @@ def meta_bernoulli(self, *, generator=None, out):
     return out
 
 
+def device_hint(tensor) -> "str":
+    if isinstance(tensor, torch._subclasses.FakeTensor):
+        return tensor.fake_device.type
+    else:
+        return "cuda"  # default to cuda
```
Comment (ezyang) [inline torch/_meta_registrations.py:316]: It'll be good to have a comment explaining what this does.

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -2895,12 +2897,25 @@ def gen_kwarg(k, v):
...
     @classmethod
     def create(cls, kernel, *args, **kwargs):
-        (
-            example_output,
-            tensor_args,
-            non_tensor_args,
-            unflatten_args,
-        ) = cls.process_kernel(kernel, *args, **kwargs)
+        fake_incorrect_kernels = (
+            aten._fft_r2c.default,
```
Comment (XiaobingSuper) [inline torch/_inductor/ir.py:2901]: @eellison, may I know why there have such fake_incorrect_kernels? I meet an issue when running a  model with a dynamic shape(the model has ```aten._fft_c2c.default```).

Comment (eellison) [reply torch/_inductor/ir.py:2901]: we need to fix it. cc @nkaretnikov who was working on updating these kernels. maybe some of these are fixed now ?

Comment (nkaretnikov) [reply torch/_inductor/ir.py:2901]: @eellison I cannot say for sure right now. Would you prefer that I take a look at these?

@XiaobingSuper Do you have a repro or more details for issues with dynamic?

Comment (nkaretnikov) [reply torch/_inductor/ir.py:2901]: FWIW I've also just remembered that there might be issues with the conj bit in fft. It's not modeled correctly on the Python side. I general fix is tricky, but maybe I can split out some of the changes from here:
https://github.com/pytorch/pytorch/pull/101835

Comment (eellison) [reply torch/_inductor/ir.py:2901]: @XiaobingSuper would you provide a repro ?

Comment (XiaobingSuper) [reply torch/_inductor/ir.py:2901]: Yes, see the following case from HF GoogleFnet:

```
import torch
import torch._dynamo
torch._dynamo.config.dynamic_shapes = True
torch._dynamo.config.assume_static_by_default = False

@torch._dynamo.optimize("inductor", dynamic=True)
def fn(a):
    convert_element_type: c64[s0, 512, 768] = torch.ops.prims.convert_element_type.default(a, torch.complex64)
    _fft_c2c: c64[s0, 512, 768] = torch.ops.aten._fft_c2c.default(convert_element_type, [1, 2], 0, True)
    return _fft_c2c

inp = torch.randn(1, 512, 768)
torch._dynamo.mark_dynamic(inp, 0)
print(fn(inp)
```

there has an error:
```
  File "/home/xiaobing/pytorch-offical/torch/_inductor/graph.py", line 643, in run_node
    result = fallback_handler(n.target, add_to_fallback_set=False)(
  File "/home/xiaobing/pytorch-offical/torch/_inductor/lowering.py", line 1152, in handler
    return pytree.tree_map(
  File "/home/xiaobing/pytorch-offical/torch/_inductor/ir.py", line 3321, in create
    ) = cls.process_kernel(kernel, *args, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/ir.py", line 2711, in process_kernel
    example_args.append(ir_node_to_tensor(x, guard_shape=True))
  File "/home/xiaobing/pytorch-offical/torch/_inductor/ir.py", line 200, in ir_node_to_tensor
    t = torch.empty_strided(
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
RuntimeError: /home/xiaobing/pytorch-offical/build/aten/src/ATen/RegisterCPU.cpp:5164: SymIntArrayRef expected to contain only concrete integers
```

Comment (XiaobingSuper) [reply torch/_inductor/ir.py:2901]: @nkaretnikov @eellison, a repro test case is given, please help see it.

Comment (nkaretnikov) [reply torch/_inductor/ir.py:2901]: Will take a look at this issue this week.

Comment (nkaretnikov) [reply torch/_inductor/ir.py:2901]: Haven't looked too deep, but I do know that the repro passes for me if I remove `c2c` from `fake_incorrect_kernels`. So I've put up a draft PR removing all fft and svd ops to see if it breaks anything: https://github.com/pytorch/pytorch/pull/103616

**Pull request closed without merging**